File: Classes/DozukiSelectSiteViewController.m
//
//  DozukiSelectSiteViewController.m
//  iFixit
//
//  Created by David Patierno on 8/16/11.
//  Copyright (c) 2011 iFixit. All rights reserved.
//

#import "DozukiSelectSiteViewController.h"
#import "iFixitAppDelegate.h"
#import "Config.h"
#import "iFixitAPI.h"
#import "UIColor+Hex.h"

#define SITES_REQUEST_LIMIT 500

static NSMutableArray *sites = nil;
static NSMutableArray *prioritySites = nil;

// Dismiss the keyboard when searchBar resigns first responder.
@implementation UIViewController (DismissKeyboard)
- (BOOL)disablesAutomaticKeyboardDismissal { return NO; }
@end

@implementation DozukiSelectSiteViewController

@synthesize searchBar, searchResults;
@synthesize simple;

- (NSString*)storedListPath {
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *docDirectory = [paths objectAtIndex:0];
    return [docDirectory stringByAppendingPathComponent:@"dozukiSiteList.plist"];
}

- (void)loadMore {
    if (!loading) {
        [sites removeAllObjects];
        [prioritySites removeAllObjects];
        loading = YES;
        [self showLoading];
        [[iFixitAPI sharedInstance] getSitesWithLimit:SITES_REQUEST_LIMIT
                                            andOffset:[sites count]
                                            forObject:self
                                         withSelector:@selector(gotSites:)];
    }
}

- (id)initWithSimple:(BOOL)simple_ {
    if ((self = [super initWithNibName:nil bundle:nil])) {
        // Custom initialization
        if (!sites) {
            sites = [[NSMutableArray array] retain];
            prioritySites = [[NSMutableArray array] retain];
        }
        hasMoreSites = YES;
        simple = simple_;
        self.title = @"Choose a Site";
        [self loadMore];
    }
    return self;
}

- (void)showLoading {
    loading = YES;
    UIView *container = [[UIView alloc] initWithFrame:CGRectMake(0.0f, 0.0f, 28.0f, 20.0f)];
    UIActivityIndicatorView *spinner = [[UIActivityIndicatorView alloc] initWithFrame:CGRectMake(0.0f, 0.0f, 20.0f, 20.0f)];
    spinner.activityIndicatorViewStyle = ([UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad) ?
        UIActivityIndicatorViewStyleGray : UIActivityIndicatorViewStyleWhite;
    [container addSubview:spinner];
    [spinner startAnimating];
    [spinner release];
    
    UIBarButtonItem *button = [[UIBarButtonItem alloc] initWithCustomView:container];
    self.navigationItem.rightBarButtonItem = button;
    [container release];
    [button release];
}
- (void)hideLoading {
    loading = NO;
    self.navigationItem.rightBarButtonItem = nil;
}
- (void)gotSites:(NSArray *)theSites {
    [self hideLoading];
    
    int count = [theSites count];
    
    if (theSites && count) {
        hasMoreSites = [theSites count] == SITES_REQUEST_LIMIT;

        // Insert these new rows at the bottom.
        NSMutableArray *paths = [NSMutableArray array];
        for (int i=0; i<[theSites count]; i++) {
            
            [paths addObject:[NSIndexPath indexPathForRow:(i + count) inSection:0]];
            
            // Check for priority sites and separate them off
            NSDictionary *site = [theSites objectAtIndex:i];
            if ([site objectForKey:@"priority"] && ![site objectForKey:@"hideFromiOS"]) {
                [prioritySites addObject:site];
            }
        }
        
        // Populate the non-priority sites list.
        for (NSDictionary *site in theSites) {
            if (![site objectForKey:@"priority"] && ![site objectForKey:@"hideFromiOS"]) {
                [sites addObject:site];
            }
        }

        [self.tableView reloadData];

        // Cache to disk.
        NSDictionary* dict = [NSDictionary dictionaryWithObjectsAndKeys:
                              prioritySites, @"prioritySites",
                              sites, @"sites",
                              nil];
        [dict writeToFile:[self storedListPath] atomically:YES];
    }
    else {
        hasMoreSites = NO;
        if ([sites count])
            return;

        // If we failed to get fresh data, use the cached site list if available.
        NSDictionary* dict = [NSDictionary dictionaryWithContentsOfFile:[self storedListPath]];
        if (![sites count] && dict) {
            [sites release];
            [prioritySites release];
            sites = [[dict objectForKey:@"sites"] mutableCopy];
            prioritySites = [[dict objectForKey:@"prioritySites"] mutableCopy];
            [self.tableView reloadData];
        }
        else {
            UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Could not load site list"
                                                            message:@"Please check your internet connection and try again."
                                                           delegate:self
                                                  cancelButtonTitle:@"Cancel"
                                                  otherButtonTitles:@"Retry", nil];
            [alert show];
            [alert release];
        }
    }
}
- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex {
    if (buttonIndex)
        [self loadMore];
}

#pragma mark - View lifecycle

- (void)viewWillAppear:(BOOL)animated {
    if (![sites count])
        [self loadMore];
}

- (void)viewDidLoad {
    [super viewDidLoad];

    // Add the search bar
    searchBar = [[UISearchBar alloc] initWithFrame:CGRectMake(0.0, 0.0, 320.0, 36.0)];
    searchBar.placeholder = @"Search";
    searchBar.delegate = self;
    self.tableView.tableHeaderView = searchBar;
    searchBar.autocorrectionType = UITextAutocorrectionTypeNo;
    
    // Update loading display status.
    if (loading)
        [self showLoading];
}

- (void)searchBarTextDidBeginEditing:(UISearchBar *)theSearchBar {
    [searchBar setShowsCancelButton:YES animated:YES];  
    
    // Animate the table up.
    if ([UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPhone) {
        [UIView beginAnimations:@"showSearch" context:nil];
        [UIView setAnimationDuration:0.3];
        CGRect bounds = self.navigationController.view.bounds;
        bounds.origin.y = 44;
        self.navigationController.view.bounds = bounds;
        [UIView commitAnimations];
    }
}

- (void)searchBarTextDidEndEditing:(UISearchBar *)theSearchBar {
    if ([theSearchBar.text isEqual:@""]) {
        searching = NO;
        noResults = NO;
        [self.tableView reloadData];
    }
    
    [searchBar setShowsCancelButton:NO animated:YES];    
    
    // Animate the table back down.
    if ([UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPhone) {
        [UIView beginAnimations:@"showSearch" context:nil];
        [UIView setAnimationDuration:0.3];
        CGRect bounds = self.navigationController.view.bounds;
        bounds.origin.y = 0;
        self.navigationController.view.bounds = bounds;
        [UIView commitAnimations];  
    }
}

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
    if ([searchText isEqual:@""]) {
        searching = NO;
        noResults = NO;
        [self.tableView reloadData];    
        self.searchResults = [NSArray array];
        return;
    }
    
    if (!searching) {
        searching = YES;
        [self.tableView reloadData];    
    }

    // Do the search in-memory.
    self.searchResults = [NSMutableArray array];
    NSMutableArray *combinedSites = [NSMutableArray arrayWithArray:prioritySites];
    [combinedSites addObjectsFromArray:sites];
    
    for (NSDictionary *site in combinedSites) {
        NSRange range = [[site objectForKey:@"title"] rangeOfString:searchText options:NSCaseInsensitiveSearch];
        if (range.location != NSNotFound) {
            [searchResults addObject:site];
        }
        else {
            range = [[site objectForKey:@"description"] rangeOfString:searchText options:NSCaseInsensitiveSearch];
            if (range.location != NSNotFound) {
                [searchResults addObject:site];
            }
        }
    }

    noResults = [searchResults count] == 0;
    
    [self.tableView reloadData];
}

- (void)searchBarCancelButtonClicked:(UISearchBar *)theSearchBar {
    searchBar.text = @"";
    noResults = NO;
    [self.view endEditing:YES];
}

- (void)searchBarSearchButtonClicked:(UISearchBar *)theSearchBar {
    [self.view endEditing:YES];
}

- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    [searchBar setShowsCancelButton:NO animated:NO];
    CGRect bounds = self.navigationController.view.bounds;
    bounds.origin.y = 0;
    self.navigationController.view.bounds = bounds;
    
    [self.view endEditing:YES];
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation {
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad)
        return YES;
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (searching) {
        if ([searchResults count])
            return [searchResults count];
        else if (noResults)
            return 1;
        else 
            return 0;
    }
    
    if (simple && [prioritySites count])
        return [prioritySites count] + 1;
    
    return [sites count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"Cell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:CellIdentifier] autorelease];
    }
    
    if (searching) {
        if ([searchResults count]) {
            cell.textLabel.text = [[searchResults objectAtIndex:indexPath.row] objectForKey:@"title"];
            cell.detailTextLabel.text = [[searchResults objectAtIndex:indexPath.row] valueForKey:@"description"];
        }
        else {
            cell.textLabel.text = @"No Results Found";
            cell.detailTextLabel.text = nil;
        }
        cell.accessoryType = UITableViewCellAccessoryNone;
    }
    else {
        NSArray *sitesArray = simple ? prioritySites : sites;
        
        // Configure the cell...
        if (simple && indexPath.row == [sitesArray count]) {
            cell.textLabel.text = @"More Sites...";
            cell.detailTextLabel.text = nil;
            cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
        }
        else {
            cell.textLabel.text = [[sitesArray objectAtIndex:indexPath.row] valueForKey:@"title"];
            cell.detailTextLabel.text = [[sitesArray objectAtIndex:indexPath.row] valueForKey:@"description"];
            cell.accessoryType = UITableViewCellAccessoryNone;
        }
    }
    
    return cell;
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
    if (searching && ![searchResults count])
        return;
    
    NSArray *sitesArray = nil;
    if (searching)
        sitesArray = searchResults;
    else if (simple)
        sitesArray = prioritySites;
    else
        sitesArray = sites;

    [TestFlight passCheckpoint:@"Dozuki Site Select"];

    if (!searching && simple && indexPath.row == [sitesArray count]) {
        DozukiSelectSiteViewController *vc = [[DozukiSelectSiteViewController alloc] initWithSimple:NO];
        [self.navigationController pushViewController:vc animated:YES];
        [vc release];
    }
    else {
        NSDictionary *site = [sitesArray objectAtIndex:indexPath.row];
        
        iFixitAppDelegate *appDelegate = (iFixitAppDelegate*)[[UIApplication sharedApplication] delegate];
        [appDelegate loadSite:site];
    }
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
    // Load the next batch if we're near the bottom.
    if ((indexPath.row >= (NSInteger)[sites count] - 1) && hasMoreSites && !loading)
        [self loadMore];
}

- (void)dealloc {
    [searchBar release];
    [searchResults release];
    [super dealloc];
}
- (void)viewDidUnload {
    [self setSearchBar:nil];
    [super viewDidUnload];
}
@end

File: iFixit/Classes/DozukiSelectSiteViewController.swift
//
//  DozukiSelectSiteViewController.m
//  iFixit
//
//  Created by David Patierno on 8/16/11.
//  Copyright (c) 2011 iFixit. All rights reserved.
//

class DozukiSelectSiteViewController: UITableViewController, UIAlertViewDelegate, UISearchBarDelegate {
    
    let SITES_REQUEST_LIMIT = 500
    
    var searchBar: UISearchBar!
    var searchResults: [Site]?
    var simple = false
    
    var loading = false
    var hasMoreSites = false
    var searching = false
    var noResults = false
    
    // Dismiss the keyboard when searchBar resigns first responder.
    override func disablesAutomaticKeyboardDismissal() -> Bool {
        return false
    }

    func storedListPath() -> NSURL {
        let urls = NSFileManager.defaultManager().URLsForDirectory(.DocumentDirectory, inDomains: .UserDomainMask)
        let url = urls[urls.count-1]
        
        return url.URLByAppendingPathComponent("dozukiSitePlist.plist")
}

    func loadMore() {
        if (!loading) {
            let config = Config.currentConfig()
            
            config.sites = []
            config.prioritySites = []

            loading = true
            self.showLoading()
            
            iFixitAPI.sharedInstance.getSites(SITES_REQUEST_LIMIT, offset:config.sites.count,
                handler:{ (results) in self.gotSites(results) })
        }
    }

    init(simple:Bool) {

        super.init(nibName:nil, bundle:nil)
        
        // Custom initialization
        hasMoreSites = true
        self.simple = simple
        self.title = NSLocalizedString("Choose a Site", comment:"")
        self.loadMore()
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder:aDecoder)
    }

    func showLoading() {
        loading = true
        let container = UIView(frame:CGRectMake(0.0, 0.0, 28.0, 20.0))
        let spinner = UIActivityIndicatorView(frame:CGRectMake(0.0, 0.0, 20.0, 20.0))
        spinner.activityIndicatorViewStyle = UIDevice.currentDevice().userInterfaceIdiom == .Pad ? .Gray : .White
        container.addSubview(spinner)
        spinner.startAnimating()
        
        let button = UIBarButtonItem(customView:container)
        self.navigationItem.rightBarButtonItem = button
    }
    
    func hideLoading() {
        loading = false
        self.navigationItem.rightBarButtonItem = nil
    }
    
    func gotSites(theSites:[Site]?) {
        let config = Config.currentConfig()

        self.hideLoading()
        
        let count = theSites?.count
        
        if (theSites != nil && count != 0) {
            hasMoreSites = count! == SITES_REQUEST_LIMIT
            
            // Insert these new rows at the bottom.
            var paths: [NSIndexPath] = []
            for i in 0 ..< count {
                paths.append(NSIndexPath(forRow: (i + count), inSection: 0))
                
                // Check for priority sites and separate them off
                let site = theSites![i]
                if (site.priority && !site.hideFromiOS) {
                    config.prioritySites.append(site)
                }
            }
            
            // Populate the non-priority sites list.
            for site in theSites! {
                if (site.priority == false && site.hideFromiOS == false) {
                    config.sites.append(site)
                }
            }
            
            self.tableView.reloadData()
            
            // Cache to disk.
            let  dict:[String:AnyObject] = ["prioritySites": config.prioritySites,
                "sites" : config.sites]

 // TODO           dict.writeToFile(self.storedListPath(), atomically:true)
        } else {
            hasMoreSites = false
            if config.sites.count != 0 {
                return
            }
            
            // If we failed to get fresh data, use the cached site list if available.
            let  dict = NSDictionary(contentsOfURL:self.storedListPath())
            if config.sites.count == 0 && dict != nil {
                config.sites = dict?["sites"] as! [Site]
                config.prioritySites = dict?["prioritySites"] as! [Site]
                self.tableView.reloadData()
            } else {
                let alert = UIAlertView(title:NSLocalizedString("Could not load site list.", comment:""),
                    message:NSLocalizedString("Please check your internet connection and try again.", comment:""),
                    delegate:self,
                    cancelButtonTitle:NSLocalizedString("Cancel", comment:""),
                    otherButtonTitles:NSLocalizedString("Retry", comment:""))
                alert.show()
            }
        }
    }
    
    func alertView(alertView:UIAlertView, clickedButtonAtIndex buttonIndex:Int) {
        if (buttonIndex != 0) {
            self.loadMore()
        }
    }

    // MARK: - View lifecycle

    override func viewWillAppear(animated:Bool) {
        let config = Config.currentConfig()

        if config.sites.count == 0 {
           self.loadMore()
        }
    }

    override func viewDidLoad() {
        let config = Config.currentConfig()
        
        super.viewDidLoad()
        
        let app = UIApplication.sharedApplication()
        
        app.statusBarHidden = false
        
        if (config.site == .Magnolia) {
            app.statusBarStyle = .Default
        } else {
            app.statusBarStyle = .LightContent
        }
        
        // Add the search bar
        searchBar = UISearchBar(frame:CGRectMake(0.0, 0.0, 320.0, 36.0))
        searchBar.placeholder = NSLocalizedString("Search", comment:"")
        searchBar.delegate = self
        self.tableView.tableHeaderView = searchBar
        searchBar.autocorrectionType = .No
        
        // For iOS 7 we want a different color scheme because the default colors are too aggressive
        UINavigationBar.appearance().barTintColor = config.toolbarColor
        self.navigationController!.navigationBar.translucent = false
        UINavigationBar.appearance().titleTextAttributes =
        [ NSForegroundColorAttributeName : config.textColor ]
        
        UINavigationBar.appearance().tintColor = config.buttonColor
        
        self.navigationItem.leftBarButtonItem!.tintColor = config.buttonColor
        self.navigationItem.rightBarButtonItem!.tintColor = config.buttonColor
        
        self.navigationItem.backBarButtonItem = UIBarButtonItem(title:"Back",
        style:.Plain, target:nil, action:nil)
        
        // Update loading display status.
        if (loading) {
            self.showLoading()
        }
    }

    func searchBarTextDidBeginEditing(theSearchBar:UISearchBar) {
        searchBar.setShowsCancelButton(true, animated:true)
    }

    func searchBarTextDidEndEditing(theSearchBar:UISearchBar) {
        if theSearchBar.text == "" {
            searching = false
            noResults = false
            self.tableView.reloadData()
        }
        
        searchBar.setShowsCancelButton(false, animated:true)
        
        // Animate the table back down.
        if UIDevice.currentDevice().userInterfaceIdiom == .Phone {
            UIView.beginAnimations("showSearch", context:nil)
            UIView.setAnimationDuration(0.3)
            var bounds = self.navigationController!.view.bounds
            bounds.origin.y = 0.0
            self.navigationController!.view.bounds = bounds
            UIView.commitAnimations()
        }
    }

    func searchBar(searchBar:UISearchBar, textDidChange searchText:String) {
        let config = Config.currentConfig()

        if searchText == "" {
            searching = false
            noResults = false
            self.tableView.reloadData()
            self.searchResults = []
            return
        }
        
        if (!searching) {
            searching = true
            self.tableView.reloadData()
        }
        
        // Do the search in-memory.
        self.searchResults = []
        let combinedSites = config.prioritySites + config.sites
        
        for site in combinedSites {
            var range = site.title?.rangeOfString(searchText, options:.CaseInsensitiveSearch)
            if range != nil {
                searchResults?.append(site)
            } else {
                range = site.siteDescription?.rangeOfString(searchText, options:.CaseInsensitiveSearch)
                if range != nil {
                    searchResults?.append(site)
                }
            }
        }
        
        noResults = searchResults?.count == 0
        
        self.tableView.reloadData()
    }

    func searchBarCancelButtonClicked(theSearchBar:UISearchBar) {
        searchBar.text = ""
        noResults = false
        self.view.endEditing(true)
    }

    func searchBarSearchButtonClicked(theSearchBar:UISearchBar) {
        self.view.endEditing(true)
    }

    override func scrollViewDidScroll(scrollView:UIScrollView) {
         searchBar.setShowsCancelButton(false, animated:false)
        var bounds = self.navigationController!.view.bounds
        bounds.origin.y = 0.0
        self.navigationController!.view.bounds = bounds
        
        self.view.endEditing(true)
    }

    // MARK: - Table view data source

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let config = Config.currentConfig()
        
        if (searching) {
            if (searchResults?.count != 0) {
                return (searchResults?.count)!
            } else if (noResults) {
                return 1
            } else {
                return 0
            }
        }
        
        if (simple && config.prioritySites.count != 0) {
            return config.prioritySites.count + 1
        }
        
        return config.sites.count
    }

    override func tableView(tableView:UITableView, cellForRowAtIndexPath indexPath:NSIndexPath) -> UITableViewCell {
        let config = Config.currentConfig()
        let CellIdentifier = "Cell"
        
        var cell = tableView.dequeueReusableCellWithIdentifier(CellIdentifier)
        if (cell == nil) {
            cell = UITableViewCell(style:.Subtitle, reuseIdentifier:CellIdentifier)
        }
        
        if (searching) {
            if searchResults!.count != 0 {
                let results = searchResults![indexPath.row]
                cell?.textLabel?.text = results.title
                cell?.detailTextLabel?.text = results.siteDescription
            } else {
                cell?.textLabel?.text = NSLocalizedString("No Results Found", comment:"")
                cell?.detailTextLabel?.text = nil
            }
            cell?.accessoryType = .None
        } else {
            let sitesArray = simple ? config.prioritySites : config.sites
            
            // Configure the cell...
            if (simple && indexPath.row == sitesArray.count) {
                cell?.textLabel?.text = NSLocalizedString("More Sites...", comment:"")
                cell?.detailTextLabel?.text = nil
                cell?.accessoryType = .DisclosureIndicator
            }
            else {
                let site = sitesArray[indexPath.row]
                cell?.textLabel?.text = site.title
                cell?.detailTextLabel?.text = site.siteDescription
                cell?.accessoryType = .None
            }
        }
        
        return cell!
    }

    // MARK: - Table view delegate

    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let config = Config.currentConfig()

        self.tableView.deselectRowAtIndexPath(indexPath, animated:true)
        
        if (searching && searchResults?.count == 0) {
            return
        }
        
        var sitesArray:[Site]!
        if (searching) {
            sitesArray = searchResults
        } else if (simple) {
            sitesArray = config.prioritySites
        } else {
            sitesArray = config.sites
        }
        
        if (!searching && simple && indexPath.row == sitesArray.count) {
            let vc = DozukiSelectSiteViewController(simple:false)
            self.navigationController!.pushViewController(vc, animated:true)
        } else {
            let site = sitesArray[indexPath.row]
            
            let appDelegate = UIApplication.sharedApplication().delegate as! iFixitAppDelegate
            appDelegate.loadSite(site)
        }
    }

    override func tableView(tableView: UITableView, willDisplayCell cell: UITableViewCell, forRowAtIndexPath indexPath: NSIndexPath) {
        let config = Config.currentConfig()
        
        // Load the next batch if we're near the bottom.
        if indexPath.row >= config.sites.count - 1 && hasMoreSites && !loading {
            self.loadMore()
        }
    }

}
